### PR TITLE
#163518703 Like Comment 

### PR DESCRIPTION
**Pull request by YvesIraguha**
#### What does this PR do?
- It helps the logged in user to like a specific comment
#### Description of Task to be completed?
- Write the tests for liking a comment on an article
- Fetch users who liked a comment

#### How should this be manually tested?
- After cloning the repository, check out `ft-like-comment-163518703` branch
- Make sure to read `README.md` file to get the setup instructions
- Using any `API client tool(usually POSTMAN)`,
- Create the user and login to get the `access-token`
- Create an article using `POST /api/v1/articles` endpoint
- Comment on the created article using `POST /api/v1/articles/:slug/comments` endpoint
- Make sure to pass the token under `request headers`. It should be {`Authorization`: `Bearer <access-token>`} KEY, VALUE PAIR
- The request's body should be as mentioned in the `README.md` file under `Endpoints:` ==> `Add Comments to an Article` subtitle
- After the successful comment creation, you should get back comment information. 
- Grab the comment Id and it use to consume `POST  /api/v1/articles/:slug/comments/<commentId>likes`. You shall see a `Comment liked message response` with the liked comment. If you try to like a comment twice, with the same access token, you shall see a `Comment unliked message` with the decremented amount of likes on a comment. 
- Consume the `GET  /api/v1/articles/:slug/comments/<commentId>likes` endpoint. You shall get a comment with the `like` and `likedBy` attributes. 

#### What are the relevant pivotal tracker stories?
- #63518703

### CODE REVIEW

File: migrations/20190220083506-add-comment-like-column.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+"use strict";
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return Promise.all([
+      queryInterface.addColumn("comments", "like", {
+        type: Sequelize.INTEGER,
```
Comment (abayo-luc) [inline migrations/20190220083506-add-comment-like-column.js:7]: What does this column do, if it is counting a number of likes, there is a specific query for that.

File: models/Comment.js [hunk 2]
```diff
@@ -26,6 +28,10 @@ export default (sequelize, DataTypes) => {
       onDelete: "CASCADE",
       hooks: true
     });
+
+    Comment.hasMany(models.commentlike, { as: "likes", foreignKey: "commentId" });
+    // associations can be defined here
```
Comment (abayo-luc) [inline models/Comment.js]: Remove this comment

File: models/User.js [hunk 3]
```diff
@@ -72,6 +72,11 @@ export default (sequelize, DataTypes) => {
       onDelete: "CASCADE",
       hooks: true
     });
+    User.hasMany(models.commentlike, {
+      as: "likedBy",
+      foreignKey: "userId"
+    });
+
     // associations can be defined here
```
Comment (abayo-luc) [inline models/User.js]: remove this comment

File: models/commentlike.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        type: DataTypes.UUID,
+        defaultValue: DataTypes.UUIDV4
+      },
+      commentId: DataTypes.UUID,
+      userId: DataTypes.UUID
+    },
+    {}
+  );
+  Commentlike.associate = function(models) {
+    // associations can be defined here
```
Comment (abayo-luc) [inline models/commentlike.js]: remove this comment

File: tests/integration/likecomment.test.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                        done(err || undefined);
+                      });
+                  });
+              }
+            });
+        }
+      });
+  });
+
+  describe("POST /articles/:slug/comments/:commentId", () => {
```
Comment (abayo-luc) [inline tests/integration/likecomment.test.js:63]: Comment description shouldn't be an endpoint.

File: tests/integration/likecomment.test.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    chai
+      .request(app)
+      .post(
+        `/api/v1/articles/${slug}/comments/192c2ff1-5622-419f-aa4b-6a4d7ad89dd4/likes`
+      )
+      .set("Authorization", `Bearer ${token}`)
+      .end((error, res) => {
+        if (error) done(error);
+        expect(res.status).equals(BAD_REQUEST);
+        expect(res.body.message).to.contain("You are liking a non-existing comment");
```
Comment (abayo-luc) [inline tests/integration/likecomment.test.js:104]: How about ```Article not found``` as an error message. (just a suggestion)

File: controllers/commentController.js [hunk 7]
```diff
@@ -2,7 +2,7 @@ import Joi from "joi";
 import models from "../models";
 import constants from "../helpers/constants";
 
-const { User, Article, Comment } = models;
+const { User, Article, Comment, commentlike } = models;
```
Comment (abayo-luc) [inline controllers/commentController.js]: We are using PascalCase, so it should be ```CommentLike```

File: controllers/commentController.js [hunk 8]
```diff
@@ -50,4 +51,99 @@ export default class CommentController {
...
+
+  static async like(req, res) {
+    const { commentId } = req.params;
+    try {
+      const comments = await Comment.findOne({ where: { id: commentId } });
+      if (!comments) {
+        throw new Error("You are liking a non-existing comment");
+      }
+    } catch (error) {
+      if (error.message == "You are liking a non-existing comment") {
```
Comment (abayo-luc) [inline controllers/commentController.js]: use ```===``` instead of ```==```

File: controllers/commentController.js [hunk 9]
```diff
@@ -50,4 +51,99 @@ export default class CommentController {
...
+
+  /**
+   * @description It helps to get all users who liked a specific comment.
+   *
+   * @param  {Object} req - The request object.
+   * @param  {Object} res - The response object.
+   * @returns {Object} - Returns the comment object.
+   */
+
+  static async getLikingUsers(req, res) {
```
Comment (abayo-luc) [inline controllers/commentController.js]: getLikingUsers is a little bit confusing,

File: controllers/commentController.js [hunk 10]
```diff
@@ -50,4 +51,99 @@ export default class CommentController {
...
+      });
+      if (comment) {
+        comment = comment.toJSON();
+        comment.likedBy = [];
+        const likes = await commentlike.findAll({
+          where: { commentId: comment.id },
+          include: [{ model: User, as: "likedBy", attributes: ["username"] }]
+        });
+
+        likes.map(like => {
```
Comment (abayo-luc) [inline controllers/commentController.js]: Can you use queries with an association http://docs.sequelizejs.com/manual/tutorial/models-usage.html

Comment (abayo-luc) [reply controllers/commentController.js]: I think that way of using ```map``` wouldn't be scalable.

**Review by abayo-luc — CHANGES_REQUESTED** [review]

File: controllers/commentController.js [hunk 11]
```diff
@@ -50,4 +51,99 @@ export default class CommentController {
       return res.status(500).send({ message: error, status: 500 });
```
Comment (niomwungeri-fabrice) [inline controllers/commentController.js]: @YvesIraguha use constans status

File: controllers/commentController.js (hunk 8 above)
Comment (niomwungeri-fabrice) [reply controllers/commentController.js]: He should activate eslint locally for automatic checking of those kinds of errors.

File: controllers/commentController.js [hunk 12]
```diff
@@ -50,4 +51,99 @@ export default class CommentController {
...
+    const { commentId } = req.params;
+    try {
+      const comments = await Comment.findOne({ where: { id: commentId } });
+      if (!comments) {
+        throw new Error("You are liking a non-existing comment");
+      }
+    } catch (error) {
+      if (error.message == "You are liking a non-existing comment") {
+        return res
+          .status(400)
```
Comment (niomwungeri-fabrice) [inline controllers/commentController.js:76]: use constants status for the rest of the file

**Review by niomwungeri-fabrice — CHANGES_REQUESTED** [review]
Please consider comments made

**Review by niomwungeri-fabrice — APPROVED** [review]

**Review by abayo-luc — APPROVED** [review]

**Review by puritybirir — APPROVED** [review]

**Pull request merged**